src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -5066,18 +5071,107 @@ BuildWaitEventSet(List *sessionList)
 			continue;
 		}
 
-		int waitEventSetIndex = AddWaitEventToSet(waitEventSet, connection->waitFlags,
-												  sock, NULL, (void *) session);
+		if (session->waitEventSetIndex == WAIT_EVENT_SET_INDEX_NOT_INITIALIZED)
+		{
+			/* should not happen, already removed from the waitEventSet */
+			continue;
+		}
+
+		int waitEventSetIndex =
+			CitusAddWaitEventSetToSet(waitEventSet, connection->waitFlags, sock,
+									  NULL, (void *) session);
 		session->waitEventSetIndex = waitEventSetIndex;
+
+		if (waitEventSetIndex == WAIT_EVENT_SET_INDEX_NOT_INITIALIZED)
+		{
+			/*
+			 * We can only lost only already connected connections,
+			 * others are regular failures.
+			 */
+			if (connection->connectionState == MULTI_CONNECTION_CONNECTED)
+			{
+				connection->connectionState = MULTI_CONNECTION_LOST;
+			}
+			else
+			{
+				connection->connectionState = MULTI_CONNECTION_FAILED;
+			}
+
+			/* immediately run the state machine to handle potential failure */
+			ConnectionStateMachine(session);
```

Comment:
Doing this in "BuildWaitEventSet" irks me a bit. 

I wonder whether we should loop through all sessions after RebuildWaitEventSet like we do at the start of RunDistributedExecution.

(Not sure)

Reply:
Hmm, I see you point. 

On the other hand, if we do not call into `ConnectionStateMachine `, then until `node_connection_timeout` the pool is not going to be considered as failed.

And, the primary reason is that `ProcessWaitEvents` would never produce events for these failed ones. Hence, none of the connections to the pool would be marked as failed.

So, maybe I should improve the comment a little like:
```
/* 
  * This connection is failed and it is our only chance to process the failure as
  * the connection would never going to generate any further events to be 
  * processed. Hence, never going to call into `ConnectionStateMachine`
  * /

Reply:
One another option could be `RebuildWaitEventSetFlags` and `BuildWaitEventSet` may output the failed connections, and we only call into `ConnectionStateMachine` for those connections.

Reply:
If we do it outside of BuildWaitEventSet then we should just do it for all sessions immediately after BuildWaitEventSet and RebuildWaitEventSetFlags. Not sure whether this is the right thing to do right now, but it does seem a bit cleaner.

Reply:
Not test yet, but why not we do it for failed epoll events like this: https://github.com/citusdata/citus/pull/5158/commits/b523d32ada3b2a50bea7d740e3fa52422debc517

We can improve the naming a bit more

Reply:
Ok, now tested by manually throwing errors, and it seems to work. Do you see any blockers to test this on the mentioned prod cluster now?